### PR TITLE
[Feature] Adds utils method remove_duplicates

### PR DESCRIPTION
**Pull request by albertbou92**
## Description

 The method detects duplicate elements in the tensor associated with the specified 'key' along the specified
'dim' and removes elements in the same indices in all other tensors within the TensorDict.

## Motivation and Context

Why is this change required? What problem does it solve?
If it fixes an open issue, please link to the issue here.
You can use the syntax `close #15213` if this solves the issue #15213

- [ ] I have raised an issue to propose this change ([required](https://github.com/pytorch/tensordict/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] I have read the [CONTRIBUTION](https://github.com/pytorch/tensordict/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: tensordict/utils.py [hunk 1]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
     if isinstance(index, slice) and (index.start == 0 or index.start is None):
         return True
     return False
+
+
+def remove_duplicates(
```
Comment (vmoens) [inline tensordict/utils.py:1867]: to be added in the doc

File: tensordict/utils.py [hunk 2]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+    tensordict: TensorDictBase, key: str, dim: int = 0
+) -> TensorDictBase:
+    """Removes indices duplicated in 'key' along the specified dimension.
+
+    This method detects duplicate elements in the tensor associated with the specified 'key' along the specified
+    'dim' and removes elements in the same indices in all other tensors within the TensorDict.
```
Comment (vmoens) [inline tensordict/utils.py:1873]:
```suggestion
    This method detects duplicate elements in the tensor associated with the specified `key` along the specified
    `dim` and removes elements in the same indices in all other tensors within the TensorDict.
```

File: tensordict/utils.py [hunk 3]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+
+    Args:
+        tensordict (TensorDictBase): The TensorDict containing potentially duplicate elements.
+        key (str): The key of the tensor along which duplicate elements should be identified and removed.
+        dim (int): The dimension along which duplicate elements should be identified and removed. Default: 0.
```
Comment (vmoens) [inline tensordict/utils.py:1878]:
```suggestion
        dim (int): The dimension along which duplicate elements should be identified and removed.
            Defaults to ``0``.
```
Also I guess dim must be within the batch-size right?

File: tensordict/utils.py (hunk 2 above)
Comment (vmoens) [reply tensordict/utils.py:1873]: We should make it clear that this all happens along the batch-size (ie, empty batch size will throw an error etc)

File: tensordict/utils.py [hunk 4]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+
+    Returns:
+        TensorDictBase: tensordict with removed elements.
```
Comment (vmoens) [inline tensordict/utils.py:1881]: tell more about the batch size of the resulting td

File: tensordict/utils.py [hunk 5]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+
+    Example:
+        >>> tensordict = TensorDict(
+        ...     {
+        ...         "tensor1": torch.tensor([[1, 2, 3], [4, 5, 6], [1, 2, 3], [7, 8, 9]]),
+        ...         "tensor2": torch.tensor([[10, 20], [30, 40], [40, 50], [50, 60]]),
+        ...     }
+        ... )
```
Comment (vmoens) [inline tensordict/utils.py:1889]: this one has no batch-size

File: tensordict/utils.py [hunk 6]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+        >>> output_tensordict = remove_duplicate_elements(tensordict, key="tensor1", dim=0)
+        >>> expected_output = TensorDict(
+        ...     {
+        ...         "tensor1": torch.tensor([[1, 2, 3], [4, 5, 6], [7, 8, 9]]),
+        ...         "tensor2": torch.tensor([[10, 20], [30, 40], [50, 60]]),
+        ...     },
+        ...     batch_size=[3],
+        ... )
+        >>> for key in output_tensordict.keys():
+        ...     assert torch.equal(output_tensordict[key], expected_output[key])
```
Comment (vmoens) [inline tensordict/utils.py:1899]:
```suggestion
        >>> assert (tensordict == expected_output).all()
```

File: tensordict/utils.py [hunk 7]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+    """
+    # Check if the key is a TensorDict
+    if key not in tensordict.keys(include_nested=True):
+        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")
+
+    # Check dim is valid
+    if dim >= len(tensordict.batch_size):
```
Comment (vmoens) [inline tensordict/utils.py:1906]: we should make it work for neg dims, like -1 (last dim of the batch size). Check other methods like unbind etc to see how to deal with that

File: tensordict/utils.py [hunk 8]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+    tensordict: TensorDictBase, key: str, dim: int = 0
+) -> TensorDictBase:
+    """Removes indices duplicated in 'key' along the specified dimension.
+
+    This method detects duplicate elements in the tensor associated with the specified 'key' along the specified
+    'dim' and removes elements in the same indices in all other tensors within the TensorDict.
+
+    Args:
+        tensordict (TensorDictBase): The TensorDict containing potentially duplicate elements.
+        key (str): The key of the tensor along which duplicate elements should be identified and removed.
```
Comment (vmoens) [inline tensordict/utils.py:1877]: We should make it extra clear that key must point to a tensor, not a tensordict (ie it must be a leaf)

Comment (vmoens) [inline tensordict/utils.py:1877]:
```suggestion
        key (NestedKey): The key of the tensor along which duplicate elements should be identified and removed.
```

File: tensordict/utils.py [hunk 9]
```diff
@@ -1862,3 +1862,65 @@ def _index_preserve_data_ptr(index):
...
+    )
+
+    # Find first occurrence of each index  (e.g. [0, 1, 3])
+    _, unique_indices_sorted = torch.sort(unique_indices, stable=True)
+    cum_sum = counts.cumsum(0)
+    cum_sum = torch.cat((torch.tensor([0]), cum_sum[:-1]))
+    first_indices = unique_indices_sorted[cum_sum]
+
+    # Remove duplicate elements in the TensorDict
+    unique_tensordict = tensordict[first_indices]
```
Comment (vmoens) [inline tensordict/utils.py:1924]: can we get the indices in the output tensordict (optionally)? like `return_indices=False` by default. The key could be `"unique_indices"`, maybe parametrizable (up to you, not sure if that makes sense). For sure we don't want to override an existing key so whatever key we choose we must check it isn't there already.

File: test/test_utils.py [hunk 10]
```diff
@@ -151,6 +152,43 @@ def test_unravel_key_to_tuple():
     assert not _unravel_key_to_tuple(("a", (slice(None),), ("b",)))
 
 
+@pytest.mark.parametrize("key", ("tensor1", "tensor3"))
+@pytest.mark.parametrize("dim", (0, 1))
```
Comment (vmoens) [inline test/test_utils.py:156]:
```suggestion
@pytest.mark.parametrize("dim", (0, 1, -1, -2))
```

File: test/test_utils.py [hunk 11]
```diff
@@ -151,6 +152,43 @@ def test_unravel_key_to_tuple():
...
+def test_remove_duplicates(key, dim):
```
Comment (vmoens) [inline test/test_utils.py:157]: I would create a td with nested keys too to make sure everything works ok

**Review by vmoens — COMMENTED** [review]
Great work! I left some suggestions to start the discussion

File: tensordict/utils.py (hunk 9 above)
Comment (albertbou92) [reply tensordict/utils.py]: We can but since the indices are an 1D tensor of length batch_size[dim], it is not guaranteed that we can add them to the tensordict. We can return them separately (optionally), which also solves the problem of potentially override an existing key.

**Comment by albertbou92** [comment]
I added the option to return the indices. 
I also handled the case of negative dims, which are now accepted.
finally I added some tests for the case of TDs of more than 1 dimension.

File: tensordict/utils.py [hunk 12]
```diff
@@ -1862,3 +1862,89 @@ def _index_preserve_data_ptr(index):
...
+    This method detects duplicate elements in the tensor associated with the specified `key` along the specified
+    `dim` and removes elements in the same indices in all other tensors within the TensorDict. It is expected for
+    `dim` to be one of the dimensions within the batch size of the input TensorDict to ensure consistency in all
+    tensors. Otherwise, an error will be raised.
+
+    Args:
+        tensordict (TensorDictBase): The TensorDict containing potentially duplicate elements.
+        key (NestedKey): The key of the tensor along which duplicate elements should be identified and removed. It
+            must be one of the leaf keys within the TensorDict, pointing to a tensor and not to another TensorDict.
+        dim (int): The dimension along which duplicate elements should be identified and removed. It must be one of
```
Comment (vmoens) [inline tensordict/utils.py:1884]: Int, optional

File: tensordict/utils.py [hunk 13]
```diff
@@ -1862,3 +1862,89 @@ def _index_preserve_data_ptr(index):
...
+    # Check if the key is a TensorDict
+    if key not in tensordict.keys(include_nested=True):
+        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")
+
+    # Check that the key points to a tensor
+    if not isinstance(tensordict.get(key), torch.Tensor):
+        raise KeyError(f"The key '{key}' does not point to a tensor in the TensorDict.")
+
+    # Check dim is valid
+    batch_dims = len(tensordict.batch_size)
```
Comment (vmoens) [inline tensordict/utils.py:1922]: Or tendordict.ndim

**Comment by albertbou92** [comment]
I added tests on different devices, turned return_indices into a kwarg only and fixed the other minor comments.

Regarding adding `unique_indices` to the output tensordict, `unique_indices` will also be a tensor of size `input.batch_size[dim]`. There is no guarantee that we can add this to the output TensorDict due to the shape (only if dim=0 and len(batch_size) = 1). This is why I decided to return it as an extra output.

**Comment by vmoens** [comment]
> I added tests on different devices, turned return_indices into a kwarg only and fixed the other minor comments.
> 
> Regarding adding `unique_indices` to the output tensordict, `unique_indices` will also be a tensor of size `input.batch_size[dim]`. There is no guarantee that we can add this to the output TensorDict due to the shape (only if dim=0 and len(batch_size) = 1). This is why I decided to return it as an extra output. is there any alternative?

Good point. The alternative would be to expand it on the left but it's messy. Happy with the PR in its current state then!

File: tensordict/utils.py [hunk 14]
```diff
@@ -1862,3 +1862,90 @@ def _index_preserve_data_ptr(index):
...
+    # Get indices of unique elements (e.g. [0, 1, 0, 2])
+    tensor = input.get(key)
+    _, unique_indices, counts = torch.unique(
+        tensor, dim=dim, sorted=True, return_inverse=True, return_counts=True
+    )
+
+    # Find first occurrence of each index  (e.g. [0, 1, 3])
+    _, unique_indices_sorted = torch.sort(unique_indices, stable=True)
+    cum_sum = counts.cumsum(0)
+    cum_sum = torch.cat((torch.tensor([0], device=input.device), cum_sum[:-1]))
```
Comment (vmoens) [inline tensordict/utils.py:1942]:
```suggestion
    cum_sum = torch.cat((torch.zeros(1, device=input.device), cum_sum[:-1]))
```

File: tensordict/utils.py [hunk 15]
```diff
@@ -1862,3 +1862,90 @@ def _index_preserve_data_ptr(index):
...
+        ... )
+        >>> assert (td == expected_output).all()
+    """
+    # Check if the key is a TensorDict
+    if key not in input.keys(include_nested=True):
+        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")
+
+    # Check that the key points to a tensor
+    if not isinstance(input.get(key), torch.Tensor):
+        raise KeyError(f"The key '{key}' does not point to a tensor in the TensorDict.")
```
Comment (vmoens) [inline tensordict/utils.py:1920]:
```suggestion
    tensor = input.get(key, default=None)

    # Check if the key is a TensorDict
    if tensor is None:
        raise KeyError(f"The key '{key}' does not exist in the TensorDict.")

    # Check that the key points to a tensor
    if not isinstance(tensor, torch.Tensor):
        raise KeyError(f"The key '{key}' does not point to a tensor in the TensorDict.")
```

File: tensordict/utils.py [hunk 16]
```diff
@@ -1862,3 +1862,90 @@ def _index_preserve_data_ptr(index):
...
+        raise ValueError(
+            f"The specified dimension '{dim}' is invalid for a TensorDict with batch size '{input.batch_size}'."
+        )
+
+    # Convert negative dimension to its positive equivalent
+    if dim < 0:
+        dim = batch_dims + dim
+
+    # Get indices of unique elements (e.g. [0, 1, 0, 2])
+    tensor = input.get(key)
```
Comment (vmoens) [inline tensordict/utils.py:1934]:
```suggestion
```

**Review by vmoens — COMMENTED** [review]
some small suggestions

**Review by vmoens — APPROVED** [review]
